src/block_url.rs

```diff
@@ -0,0 +1,140 @@
+                                success = false;
+                            }
+                            // TODO: check if url is within db, and if not, add it to database
+                        },
+                        _ => {}
+                    }
+                } else {
+                    info!("Robots.txt couldn't be parsed");
+                    success = false; // failsafe for broken robots.txt
+                }
+                if !success { return Ok(false); }
+            }
+            Ok(true)
+        },
+        Err(e) => Err(e)
+    }
+}
+
+async fn check_blocked_list(connection: &Connection, url: &str) -> Result<bool, Box<dyn Error>> {
+    // TODO: check timestamp and compare
+    let blocked_urls = get_blocked_list(connection, url).await?; 
+    let sub_path = get_subpath(url)?;
+
+    for url in blocked_urls.sub_url {
+        if url == sub_path { return Ok(false) }
+    }
+    Ok(true)
+}
+
+async fn get_blocked_list(connection: &Connection, url: &str) -> Result<BlockedList, Box<dyn Error>> {
+    let root_url = get_url_path(url, None)?;
+    let root: (i64, String) = connection.query_row(
+        "SELECT Id, Modified FROM RootUrl WHERE Url = ?1",
+        [&root_url],
+        |row| {
+            let id = row.get(0)?;
+            let modified = row.get(1)?;
+            Ok((id, modified))
+        },
+    )?;
+    
+    let blocked_list: Vec<String> = connection
+        .prepare("SELECT SubUrl FROM BlockedUrl WHERE RootId = ?")?
+        .query_map([root.0], |row| row.get::<_, String>(0))?
+        .collect::<Result<Vec<_>, _>>()?;
+    
+    Ok(BlockedList {
+        root_id: root.0,
+        sub_url: blocked_list,
+        modified: root.1,
+    })
+}
+
+fn get_subpath(url: &str) -> Result<String, Box<dyn std::error::Error>> {
+    let parsed_url = Url::parse(url)?;
+    let subpath = parsed_url.path().trim_start_matches('/');
+    Ok(subpath.to_string())
+}
+
+fn get_url_path(url: &str, path: Option<&str>) -> Result<String, Box<dyn Error>>  {
+    let local_path = match path {
+        Some(path) => path, 
+        None => "/"
+    };
+    
+    let parsed = Url::parse(url)?;
+    let mut robots_url = parsed.clone();
+    
+    robots_url.set_path(local_path);
+    robots_url.set_query(None);
+    robots_url.set_fragment(None);
+    Ok(robots_url.as_str().to_string())
+}
+
+async fn add_blocked_to_db(connection: &Connection) {
+    // TODO: add new blocked urls to db
+}
+
+async fn get_robots(client: &Client, url: &str) -> Result<String, Box<dyn Error>> {
+    let robots_url = get_url_path(url, Some("/robots.txt"))?;
+    let result = client.get(robots_url).send().await?;
+    if !result.status().is_success() {
+        return Err(Box::new(io::Error::new(io::ErrorKind::Other, "No robots.txt" ))); // failsafe for no robots.txt
+    }
+
+    Ok(result.text().await?)
+}
+
+async fn parse_robots(client: &Client, url: &str) -> Result<Vec<(Option<String>, Option<String>)>, Box<dyn Error>> {
```

Comment:
Parsing a robots file and fetching it should be separate behaviours. Please move `get_robots` out of `parse_robots` into the calling method. This will improve clarity. Comments should be added as well.

```suggestion
// Caller:
...
let robotsContent = get_robots(client, url).await
let blockedUrls = parse_robots(robotsContent).await
...

/// Fetches the robots.txt file from a domain specified in the url.
/// This is returned as a string. (...etc)
async fn get_robots(client: &Client, url: &str)-> Result<String, Box<dyn Error>> { ... }

/// Parses the contents of a robots.txt file
/// Returns a list of key-value pairs (...etc)
async fn parse_robots(robotsContent: &str) -> Result<Vec<(Option<String>, Option<String>)>, Box<dyn Error>> { ... }
```